src/server/replica.cc

```diff
@@ -431,43 +433,62 @@ error_code Replica::InitiateDflySync() {
     shard_flows_[i].reset(new Replica(master_context_, i, &service_));
   }
 
-  SyncBlock sb{num_df_flows_};
+  // Allocate shared, because the error handler might outlive the whole task.
+  auto sync_block = std::make_shared<SyncBlock>(num_df_flows_);
+
+  // Init context.
+  // From full sync and onwards cleanup is managed by the underlying currently running task.
+  // Context responsibility is always transferred atomically.
+  {
+    lock_guard lk(cntx_mu_);
+    cntx_.emplace([this, sync_block](const auto& ge) {
+      // Prevent blocking fiber that issued error by starting another fiber.
+      ::boost::fibers::fiber{[this, sync_block]() {
+        // Unblock sync block.
+        {
+          lock_guard lk(sync_block->mu_);
+          sync_block->flows_left = 0;
+        }
+        sync_block->cv_.notify_all();
+      }}.detach();
+
+      // Close sockets to unblock flows in case of cancellation.
+      CloseAllSockets();
+      return true;
+    });
+  }
```

Comment:
Okay, I've tried my best... 
The whole reason for this complexity (callbacks) is that we can't block on two things at once. 
For example, in Go we have a robust select that can read a socket, check the context, check a mutex, etc.

Here, I need to make sure that we don't block in every possible scenario. Those are: 
- if a single flow fails, then we need to wake the other flows by closing their sockets, so cancellation is fast
- If one or more flows fail, then the sync block prevents the main coordinator from proceeding, so we need to unblock it. Either by letting flows do cleanup for errors (difficult code flow), or by doing cleanup in a centralized place (seems to be easier to manage, if this functions own the resource, then it should clean up).

I'm not sure yet if we need to spawn a dedicated fiber. Either way, CloseAllSockets needs to run _inside_ the handler (forgot comment), because the sockets might be dropped afterwards.

Reply:
why you do not use already existing `fibers_ext::BlockingCounter` instead of custom `SyncBlock`?
not only it already exists, it does not require mutex locking when notifying so the code here will be much simpler (no need a fiber, lock_guard).

Reply:
I know about it. Because the BlockingCounter is _blocking_, but I need all the flows to continue processing incremental changes. I can use it, and instead spawn a new fiber that blocks it. This is also a possible solution

I guess we discussed this in one of the replication PRs